File: backend/app/models/item.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, Float, Text, DateTime, Boolean, ForeignKey
from app.database.base import Base


class Item(Base):
    __tablename__ = "items"

    id                      = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    owner_id                = Column(String(36), ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    title                   = Column(String, nullable=False)
    category                = Column(String, nullable=True)   # e.g. "Electronics", "Furniture"
    base_price              = Column(Float, nullable=False)
    current_price           = Column(Float, nullable=False)
    description             = Column(Text, nullable=True)
    image_url               = Column(String, nullable=True)
    whatsapp_number         = Column(String, nullable=False)
    auction_ends_at         = Column(DateTime, nullable=False)
    payment_method          = Column(String, nullable=False, default="Cash on Delivery")
    current_bidder_name     = Column(String, nullable=True)
    current_bidder_whatsapp = Column(String, nullable=True)
    current_bidder_id       = Column(String(36), ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    bid_updated_at          = Column(DateTime, nullable=True)
    is_finalized            = Column(Boolean, default=False)
    created_at              = Column(DateTime, default=datetime.utcnow)

File: DormDeal/lib/models/item.dart
class Item {
  final String id;
  final String? ownerId;
  final String title;
  final double basePrice;
  final double currentPrice;
  final String? description;
  final String? category;
  final String whatsappNumber;
  final String? imageUrl;
  final DateTime auctionEndsAt;
  final String paymentMethod;
  final String? currentBidderName;
  final String? currentBidderWhatsapp;
  final String? currentBidderId;
  final DateTime? bidUpdatedAt;
  final bool isFinalized;
  final int bidCount;
  final DateTime createdAt;

  Item({
    required this.id,
    this.ownerId,
    required this.title,
    required this.basePrice,
    required this.currentPrice,
    this.description,
    this.category,
    required this.whatsappNumber,
    this.imageUrl,
    required this.auctionEndsAt,
    required this.paymentMethod,
    this.currentBidderName,
    this.currentBidderWhatsapp,
    this.currentBidderId,
    this.bidUpdatedAt,
    this.isFinalized = false,
    this.bidCount = 0,
    required this.createdAt,
  });

  factory Item.fromJson(Map<String, dynamic> json) {
    return Item(
      id: json['id'] ?? '',
      ownerId: json['owner_id'],
      title: json['title'] ?? '',
      basePrice: ((json['base_price'] ?? 0) as num).toDouble(),
      currentPrice: ((json['current_price'] ?? json['base_price'] ?? 0) as num).toDouble(),
      description: json['description'],
      category: json['category'],
      whatsappNumber: json['whatsapp_number'] ?? '',
      imageUrl: json['image_url'],
      auctionEndsAt: json['auction_ends_at'] != null
          ? DateTime.parse(json['auction_ends_at'])
          : (json['created_at'] != null
              ? DateTime.parse(json['created_at']).add(const Duration(hours: 24))
              : DateTime.now().add(const Duration(hours: 24))),
      paymentMethod: json['payment_method'] ?? 'Cash on Delivery',
      currentBidderName: json['current_bidder_name'],
      currentBidderWhatsapp: json['current_bidder_whatsapp'],
      currentBidderId: json['current_bidder_id'],
      bidUpdatedAt: json['bid_updated_at'] != null
          ? DateTime.parse(json['bid_updated_at'])
          : null,
      isFinalized: json['is_finalized'] ?? false,
      bidCount: json['bid_count'] ?? 0,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : DateTime.now(),
    );
  }
}
